### PR TITLE
slof_multi_devices: Add slof_multi_devices case for testing SLOF

### PR DESCRIPTION
**Pull request by YongxueHong**
slof_multi_devices case is that verify SLOF info
with multi-devices.

ID: 1645019

Signed-off-by: Yongxue Hong <yhong@redhat.com>

### CODE REVIEW

**Comment by YongxueHong** [comment]
Depends on https://github.com/autotest/tp-qemu/pull/1531

**Comment by wangxianxian** [comment]
python3 ConfigTest.py --driveformat=virtio_blk --nicmodel=virtio_net --imageformat=qcow2 --guestname=RHEL.x.x --platform=ppc64le --mem=4096 --vcpu=4 --testcase=slof_multi_devices
slof_multi_devices.with_pci_bridge: PASS
slof_multi_devices.with_virtio_net_pci: PASS

python3 ConfigTest.py --driveformat=virtio_blk --nicmodel=spapr-vlan --imageformat=qcow2 --guestname=RHEL.x.x --platform=ppc64le --mem=4096 --vcpu=4 --testcase=slof_multi_devices --clone=no
slof_multi_devices.with_pci_bridge: PASS

File: qemu/tests/cfg/slof_multi_devices.cfg [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    pci_controllers = ''
+    only ppc64le ppc64
+    only Linux
+    # select one drive format since just test slof
+    # with pci-bridge and virtio-net-pci.
+    only virtio_blk
+    variants:
+        - with_pci_bridge:
+            device_type = 'pci-bridge'
+            pci_controllers +=" pci_bridge0"
```
Comment (wangxianxian) [inline qemu/tests/cfg/slof_multi_devices.cfg]: It does not need to assign value to "pci_controllers" and "type_pci_bridge0" here, it's better add a parameter that we could set it as our need to define the number of pci-bridge.

File: qemu/tests/slof_multi_devices.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        2. Check if any error info from output of SLOF.
+        3. Guest could login sucessfully.
+        4. Guest could ping external host ip.
+        5. For virtio-net-pci scenario, check the number of NIC if equal to qemu command.
+    :param test: Qemu test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    if params['device_type'] == 'pci-bridge':
+        for id in range(0, 2):
```
Comment (wangxianxian) [inline qemu/tests/slof_multi_devices.py]: It's better to pass a variable to function like rang(0, num), the value of "num" could be got from cfg.

File: qemu/tests/slof_multi_devices.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+     4. Guest could ping external host ip.
+     5. For virtio-net-pci scenario, check the number of NIC if equal to
+        qemu command.
+
+    :param test: Qemu test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+    if params['device_type'] == 'pci-bridge':
+        for id in range(0, int(params['pci_bridge_num'])):
```
Comment (wangxianxian) [inline qemu/tests/slof_multi_devices.py]: 'pci_bridge_num' should be added in slof_multi_devices.cfg

**Comment by wangxianxian** [comment]
Ack

**Comment by YongxueHong** [comment]
@hereischen  @lolyu  help me review, thanks.

File: qemu/tests/cfg/slof_multi_devices.cfg [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            type_pci_bridge0 = pci-bridge
```
Comment (hereischen) [inline qemu/tests/cfg/slof_multi_devices.cfg]: It seems you don't need this line

File: qemu/tests/slof_multi_devices.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    if params['device_type'] == 'pci-bridge':
+        for id in range(int(params['pci_bridge_num'])):
+            params['pci_controllers'] += ' pci_bridge%d' % id
+            params['type_pci_bridge%d' % id] = 'pci-bridge'
+    elif params['device_type'] == 'virtio-net-pci':
+        pci_num = int(params['pci_bridge_num'])
+        nic_id = 0
+        for pci_id in range(pci_num):
+            params['pci_controllers'] += ' pci_bridge%d' % pci_id
+            params['type_pci_bridge%d' % pci_id] = 'pci-bridge'
```
Comment (hereischen) [inline qemu/tests/slof_multi_devices.py]: You can put `line 35 36 37 39 41 42 43` outside in the if clauses

**Comment by hereischen** [comment]
Acked-by: Haotong Chen hachen@redhat.com

**Comment by YongxueHong** [comment]
Thanks all.

**Pull request merged**